File: companies/Apple/251.flatten-2-d-vector.java
/**
 * Algorithm:
 * Keep row and column pointers. Before next or hasNext, skip empty rows until
 * row points to an available value or reaches the end.
 *
 * Java data structures:
 * int[][] stores the input vector; row/col are cursor indices.
 *
 * Complexity:
 * next and hasNext are O(1) amortized; space O(1).
 */
class Vector2D {
    private final int[][] vec;
    private int row;
    private int col;

    public Vector2D(int[][] vec) {
        this.vec = vec;
        row = 0;
        col = 0;
    }

    public int next() {
        skipEmpty();
        return vec[row][col++];
    }

    public boolean hasNext() {
        skipEmpty();
        return row < vec.length;
    }

    private void skipEmpty() {
        while (row < vec.length && col >= vec[row].length) {
            row++;
            col = 0;
        }
    }
}

File: companies/Apple/251.flatten-2-d-vector.go
package leetcode

// Vector2D251 stores row and column indices into the original 2D slice. skipEmpty
// advances over empty rows before Next or HasNext, so both methods handle sparse
// row lengths consistently.
type Vector2D251 struct {
	vec      [][]int
	row, col int
}

func Constructor251(vec [][]int) Vector2D251 { return Vector2D251{vec: vec} }
func (v *Vector2D251) skipEmpty() {
	for v.row < len(v.vec) && v.col >= len(v.vec[v.row]) {
		v.row++
		v.col = 0
	}
}
func (v *Vector2D251) Next() int     { v.skipEmpty(); val := v.vec[v.row][v.col]; v.col++; return val }
func (v *Vector2D251) HasNext() bool { v.skipEmpty(); return v.row < len(v.vec) }
